### PR TITLE
NEXUS-22747 Conan: Database migration for 1.0.0 release

### PR DESCRIPTION
**Pull request by mlukaretkyi**
https://issues.sonatype.org/browse/NEXUS-22747

### CODE REVIEW

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import static org.sonatype.nexus.repository.storage.AssetEntityAdapter.P_ASSET_KIND;
+
+/**
+ * @since 1.0.0
+ */
+@Named
+@Singleton
+@Upgrades(model = ConanModel.NAME, from = "1.0", to = "1.1")
+@DependsOn(model = DatabaseInstanceNames.COMPONENT, version = "1.14", checkpoint = true)
+@DependsOn(model = DatabaseInstanceNames.CONFIG, version = "1.8", checkpoint = true)
```
Comment (dbradicich) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: appears there are no changes made to the config database (only the component database), so i don't think you need `checkpoint = true` here

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+      @Named(DatabaseInstanceNames.CONFIG) final Provider<DatabaseInstance> configDatabaseInstance,
+      @Named(DatabaseInstanceNames.COMPONENT) final Provider<DatabaseInstance> componentDatabaseInstance)
+  {
+    this.configDatabaseInstance = checkNotNull(configDatabaseInstance);
+    this.componentDatabaseInstance = checkNotNull(componentDatabaseInstance);
+  }
+
+  @Override
+  public void apply() {
+    if (hasSchemaClass(configDatabaseInstance, "repository") &&
```
Comment (dbradicich) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: might as well add static final string for REPOSITORY_CLASS_NAME (keep consistent as its done for ASSET_CLASS_NAME already)

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                oDocument.save();
+              });
+        }));
+  }
+
+  private void updateAssetPath(final List<String> repositoryNames) {
+    DatabaseUpgradeSupport.withDatabaseAndClass(componentDatabaseInstance, ASSET_CLASS_NAME,
+        (db, type) -> repositoryNames
+            .stream()
+            .flatMap(repositoryName -> {
```
Comment (dbradicich) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: might align this method with the `removeAttributesFromConanManifest` method, in that you could use this flatMap logic in both instances (or alternatively don't use the flatMap on the stream in this method, and simply do a forEach on repositoryNames as done in the other method, don't particularly care which, but should be consistent)

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+              List<ODocument> assets =
+                  db.query(new OSQLSynchQuery<ODocument>("select from asset where bucket = ?"), bucket.getIdentity());
+              return assets.stream();
+            })
+            .forEach(oDocument -> {
+              String name = oDocument.field("name");
+              String nextName = null;
+
+              if (name.startsWith("/v1/conans/v1/conans/")) { // broken HOSTED
+                nextName = "conans/" + name.substring(21);
```
Comment (dbradicich) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: might replace `21` with `"/v1/conans/v1/conans/".length()` (and similar in the other else ifs below, just to keep folks from trying to figure what the number relates to)

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 1 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: fixed https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/9938f7c67cac959dbbf6292c9271250ee6812827

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 2 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: fixed https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/9938f7c67cac959dbbf6292c9271250ee6812827

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 3 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: was created one method called findAssets
fixed https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/9938f7c67cac959dbbf6292c9271250ee6812827

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 4 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: not a bad idea.
https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/9938f7c67cac959dbbf6292c9271250ee6812827

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 1 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: You know, after this changes IT is failing in CI and my local env.
I got
org.sonatype.nexus.common.stateguard.InvalidStateException: Invalid state: NEW; allowed: [STARTED]

Maybe you make mistake about checkpoint=false ?
With checkpoint=true, everything ok

Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: Thanks, removed checkpoint = true from CONFIG
but checkpoint = true  in COMPONENT setted

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+              if (nextName != null) {
+                oDocument.field(P_ASSET_NAME, nextName);
+                oDocument.save();
+              }
+            })
+    );
+  }
+
+  private Stream<ODocument> findAssets(
```
Comment (proteus1121) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: Is it necessary to return stream here, List will be clearer up to me?

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 1 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/b669860e4a223502adf1793e400af5d579cf2c76

File: nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+  static final String ASSET_CLASS = new OClassNameBuilder()
+      .type("asset")
+      .build()
+
+  static final String I_ASSET_NAME = new OIndexNameBuilder()
+      .type(ASSET_CLASS)
+      .property(P_NAME)
+      .build()
+
+  static final List<String> ACTUAL_NAMES = Collections.unmodifiableList(Arrays.asList(
```
Comment (proteus1121) [inline nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy]: just ImmutableList.of (... , ...)

File: nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+          ".txt",
+      "/v1/conans/v1/conans/react/1.2.11/conan/stable/packages/534dcc368c999e07e81f146b3466b8f656ef1f55/download_urls",
+      "/v1/conans/v1/conans/conan/react/1.2.11/stable/conan_export.tgz",
+      "/v1/conans/v1/conans/conan/react/1.2.11/stable/conanfile.py",
+      "/v1/conans/v1/conans/conan/react/1.2.11/stable/conanmanifest.txt",
+      "/v1/conans/v1/conans/react/1.2.11/conan/stable/download_urls"
+      // react hosted base on v1/api-refactor testing
+  ))
+
+  static final List<String> EXPECTED_NAMES = Collections.unmodifiableList(Arrays.asList(
```
Comment (proteus1121) [inline nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy]: same as upper

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+          .stream()
+          .map(entries -> (String) entries.field(P_REPOSITORY_NAME))
+          .collect(Collectors.toList());
+    }
+  }
+
+  private void removeAttributesFromConanManifest(final List<String> repositoryNames) {
+    DatabaseUpgradeSupport.withDatabaseAndClass(componentDatabaseInstance, ASSET_CLASS_NAME,
+        (db, type) -> findAssets(db, repositoryNames,
+            "select from asset where bucket = ? and attributes.conan.asset_kind = 'CONAN_MANIFEST'")
```
Comment (proteus1121) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: AssetKind.CONAN_MANIFEST.name ?

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 5 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: I need foreach.
Yes, you are right, I can find assets and after use collect(Collectors.toList()) and use List
But, I need class which has foreach, so I can skip transformation collect(Collectors.toList()) and just use Stream

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 8 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: fixed https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/6bac95f207f3b84e6c3e28f23661ea59eb9c31fb

File: nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy (hunk 7 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy]: fixed https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/6bac95f207f3b84e6c3e28f23661ea59eb9c31fb

File: nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy (hunk 6 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/test/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1_Test.groovy]: fixed https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/6bac95f207f3b84e6c3e28f23661ea59eb9c31fb

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+          .stream()
+          .map(entries -> (String) entries.field(P_REPOSITORY_NAME))
+          .collect(Collectors.toList());
+    }
+  }
+
+  private void removeAttributesFromConanManifest(final List<String> repositoryNames) {
+    DatabaseUpgradeSupport.withDatabaseAndClass(componentDatabaseInstance, ASSET_CLASS_NAME,
+        (db, type) -> {
+          String SQL = String.format("select from asset where bucket = ? and attributes.conan.asset_kind = '%s'",
```
Comment (proteus1121) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: please rename

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+      return documents
+          .stream()
+          .map(entries -> (String) entries.field(P_REPOSITORY_NAME))
+          .collect(Collectors.toList());
+    }
+  }
+
+  private void removeAttributesFromConanManifest(final List<String> repositoryNames) {
+    DatabaseUpgradeSupport.withDatabaseAndClass(componentDatabaseInstance, ASSET_CLASS_NAME,
+        (db, type) -> {
```
Comment (chernikovmaksym) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: Consider moving lambda function to a separate method

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                oDocument.save();
+              });
+        }
+    );
+  }
+
+  private void updateAssetPath(final List<String> repositoryNames) {
+    DatabaseUpgradeSupport.withDatabaseAndClass(componentDatabaseInstance, ASSET_CLASS_NAME,
+        (db, type) -> findAssets(db, repositoryNames, "select from asset where bucket = ?")
+            .forEach(oDocument -> {
```
Comment (chernikovmaksym) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: Consider moving lambda function to a separate method

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+  }
+
+  private Stream<ODocument> findAssets(
+      final ODatabaseDocumentTx db,
+      final List<String> repositoryNames,
+      final String SQL)
+  {
+    return repositoryNames
+        .stream()
+        .flatMap(repositoryName -> {
```
Comment (chernikovmaksym) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: Consider moving lambda function to a separate method

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java (hunk 9 above)
Comment (mlukaretkyi) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/upgrade/ConanUpgrade_1_1.java]: flixed https://github.com/sonatype-nexus-community/nexus-repository-conan/pull/95/commits/c3bdbb26abe01b2d245ce41b87da169dbad8bd59

**Pull request merged**